src/utils/polygon.cpp

```diff
@@ -156,6 +156,89 @@ Polygons PolygonRef::offset(int distance, ClipperLib::JoinType joinType, double
     return ret;
 }
 
+Polygon Polygons::convexHull() const
+{
+    // Implements Andrew's monotone chain convex hull algorithm
+    // See https://en.wikibooks.org/wiki/Algorithm_Implementation/Geometry/Convex_hull/Monotone_chain
+
+    using IntPoint = ClipperLib::IntPoint;
+
+    std::vector<IntPoint> all_points;
+    for (const ClipperLib::Path &path : paths)
+    {
+        // Reserve big enough for result, but also maintain O(1)
+        // amortized resize cost
+        size_t reserve_size = std::max(all_points.size() + path.size(),
+                                       2*all_points.size());
+        all_points.reserve(reserve_size);
+        for (const IntPoint &point : path)
+        {
+            all_points.push_back(point);
+        }
+    }
+
+    struct HullSort
+    {
+        bool operator()(const IntPoint &a,
+                        const IntPoint &b)
+        {
+            return (a.X < b.X) ||
+                (a.X == b.X && a.Y < b.Y);
+        }
+    };
+
+    std::sort(all_points.begin(), all_points.end(), HullSort());
+
+    // positive for left turn, 0 for straight, negative for right turn
+    auto ccw = [](const IntPoint &p0, const IntPoint &p1, const IntPoint &p2) -> int64_t {
+        IntPoint v01(p1.X - p0.X, p1.Y - p0.Y);
+        IntPoint v12(p2.X - p1.X, p2.Y - p1.Y);
+
+        return
+            static_cast<int64_t>(v01.X) * v12.Y -
+            static_cast<int64_t>(v01.Y) * v12.X;
+    };
+
+    size_t num_points = all_points.size();
+
+    Polygon hull_poly;
+    ClipperLib::Path &hull_points = *hull_poly;
+    hull_points.resize(2*num_points);
```

Comment:
I don't think your example case produces more than "num_points + 1" points.  This convex hull implementation does not include points that are colinear on the hull due to == being in the decisions on backing out a point.  So when producing the lower hull, it would end up with the lower left point, the lower right point, and the upper right point only.  All the intermediate points on the right middle wouldn't be in the hull.  Similarly, when constructing the upper hull, it will only have the upper right point and the lowest point it has reached so far until it gets to the upper left point.  It never has more than 5 points in the convex hull I think for the case you mentioned.